librz/reg/profile.c

```diff
@@ -1,245 +1,452 @@
+ *
+ * Offset is of the form: <byte>.<bit>
+ * .<bit> is optional.
+ *
+ * \param s Offset string.
+ * \param def The defintion item to store the offset into \p def->offset in bits.
+ * \return false On failure (sets def->offset = UT32_MAX).
+ * \return true On success.
+ */
+static bool parse_offset(const char *s, RZ_OUT RzRegProfileDef *def) {
+	rz_return_val_if_fail(s && def, false);
+	if (s[0] == '?') {
+		def->offset = UT32_MAX;
+		return true;
+	} else if (s[0] == '.') {
+		def->offset = strtoul(s + 1, NULL, 0);
+		return true;
 	}
+	def->offset = strtoul(s, NULL, 0) * 8;
 
-	item->type = type;
-	item->name = strdup(tok[1]);
-	// All the numeric arguments are strictly checked
-	item->size = parse_size(tok[2], &end);
-	if (*end != '\0' || !item->size) {
-		rz_reg_item_free(item);
-		return "Invalid size";
+	const char *bi = strchr(s, '.');
+	if (!bi) {
+		// No bit offset given.
+		return true;
 	}
-	if (!strcmp(tok[3], "?")) {
-		item->offset = -1;
-	} else {
-		item->offset = parse_size(tok[3], &end);
+
+	ut8 bit_offset = strtoul(bi + 1, NULL, 0);
+	def->offset += bit_offset;
+	return true;
+}
+
+/**
+ * \brief Parses a register alias.
+ *
+ * The alias is of the form:
+ * "=<alias>  <reg name>"
+ *
+ * \param reg The RzReg struct with the register profile.
+ * \param tokens A list with both tokens of the alias string.
+ * \return true On success.
+ * \return false On Failure.
+ */
+static bool parse_alias(RZ_OUT RzReg *reg, RZ_BORROW RzList *tokens) {
+	rz_return_val_if_fail(reg && tokens, false);
+	RzRegProfileAlias *pa = RZ_NEW0(RzRegProfileAlias);
+	if (!pa) {
+		RZ_LOG_WARN("Unable to allocate memory.\n");
+		return false;
 	}
-	if (*end != '\0') {
-		rz_reg_item_free(item);
-		return "Invalid offset";
+
+	const char *real_name = rz_list_get_n(tokens, 1);
+	const char *alias = rz_list_get_n(tokens, 0);
+	if (!alias) {
+		RZ_LOG_WARN("Failed to get alias name from token.\n")
+		return false;
 	}
-	item->packed_size = parse_size(tok[4], &end);
-	if (*end != '\0') {
-		rz_reg_item_free(item);
-		return "Invalid packed size";
+
+	RzRegisterId role = rz_reg_get_name_idx(alias + 1);
+	if (!(role >= 0 && role < RZ_REG_NAME_LAST)) {
+		RZ_LOG_WARN("Invalid alias\n");
+		return false;
 	}
 
-	// Dynamically update the list of supported bit sizes
-	reg->bits |= item->size;
+	pa->alias = strdup(alias);
+	pa->reg_name = strdup(real_name);
+	pa->role = role;
+	rz_list_append(reg->reg_profile.alias, pa);
+
+	return true;
+}
+
+/**
+ * \brief Parses a register definition.
+ *
+ * \param reg Register struct with the register profile.
+ * \param tokens List of strings of a single register definition.
+ * \return false On failure.
+ * \return true On success.
+ */
+static bool parse_def(RZ_INOUT RzReg *reg, RZ_OWN RzList *tokens) {
+	rz_return_val_if_fail(reg && tokens, false);
+
+	const char *name = rz_list_get_n(tokens, 1);
+	rz_return_val_if_fail(name, false);
```

Comment:
I am not sure it should be an assert, @wargio @thestr4ng3r  @ret2libc thoughts?

Reply:
It should be a non-assert error since profiles can be user-supplied any time.

Reply:
Can even be removed. List length is checked before. So this will always result in a string.

Reply:
Then assert is fine.